src/crypto/CryptStateEVP.h

```diff
@@ -0,0 +1,287 @@
+// Copyright 2021 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+// This code aims to provide a unified interface to operate the EVP interface of OpenSSL.
+// See https://wiki.openssl.org/index.php/EVP_Authenticated_Encryption_and_Decryption
+
+#ifndef MUMBLE_CRYPTOSTATEEVP_H
+#define MUMBLE_CRYPTOSTATEEVP_H
+
+#include "CryptState.h"
+
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+
+
+template< class T > class CryptStateEVP : public CryptState {
+public:
+	explicit CryptStateEVP(const EVP_CIPHER *c);
+	~CryptStateEVP() override;
+
+	bool isValid() const override;
+	void genKey() override;
+	bool setKey(const std::string &rkey, const std::string &eiv, const std::string &div) override;
+	bool setRawKey(const std::string &rkey) override;
+	bool setEncryptIV(const std::string &iv) override;
+	bool setDecryptIV(const std::string &iv) override;
+	std::string getRawKey() override;
+	std::string getEncryptIV() override;
+	std::string getDecryptIV() override;
+
+	bool decrypt(const unsigned char *source, unsigned char *dst, unsigned int encrypted_length,
+				 unsigned int &plain_length) override;
+	bool encrypt(const unsigned char *source, unsigned char *dst, unsigned int plain_length,
+				 unsigned int &encrypted_length) override;
+
+private:
+	static constexpr unsigned int IV_LENGTH  = 0;
+	static constexpr unsigned int KEY_LENGTH = 0;
+	static constexpr unsigned int TAG_LENGTH = 0;
```

Comment:
The reason for using a Template is to override these `static const` correctly. I think it is very important to keep these variables constant in order to help the compiler optimize the encryption part of our code. These constants will be defined in each derived class and be accessed by `T::IV_LENGTH`, etc. in the following code.